src/write_gcm.c

```diff
@@ -3737,12 +3872,60 @@ static int wg_rebase_item(c_avl_tree_t *deriv_tree, wg_payload_t *payload,
   return 0;
 }
 
+static c_avl_tree_t *wg_group_payloads_by_host(wg_payload_t *head, int *count) {
+  c_avl_tree_t *host_tree = c_avl_create((
+      int (*)(const void*, const void*))&strcmp);
+  if (host_tree == NULL) return NULL;
+  int host_count = 0;
+  while (head != NULL) {
+    const char *host = head->key.host;
+    wg_payload_hook_t *hook = NULL;
+    wg_payload_hook_t *value = NULL;
+    if (c_avl_get(host_tree, (const void *) host, (void **) &value) != 0) {
+      hook = (wg_payload_hook_t *) calloc(1, sizeof(wg_payload_hook_t));
+      if (hook == NULL) {
+        ERROR("write_gcm: wg_group_by_host malloc failed");
+        goto leave;
+      }
+      hook->payload = head;
+      if (c_avl_insert(host_tree, (void *) host, (void *) hook) < 0) {
+        ERROR("write_gcm: wg_group_by_host tree insert failed");
+        goto leave;
+      }
+      wg_payload_t *next = head->next;
+      head->next = NULL;
+      head = next;
+      ++host_count;
+    } else {
+      if (value == NULL) {
+        ERROR("write_gcm: wg_group_by_host tree insert failed");
+        goto leave;
+      }
+      hook = value;
+      wg_payload_t *next = head->next;
+      head->next = hook->payload;
+      hook->payload = head;
+      head = next;
+    }
+  }
+  *count = host_count;
+  return host_tree;
+ leave:
+  *count = -1;
+  wg_payload_destroy(head);
```

Comment:
Why release the memory here instead of having the caller be responsible for this?